lib/runner.js

```diff
@@ -52,11 +58,45 @@ class Runner extends Emittery {
 			if (!scheduledStart) {
 				scheduledStart = true;
 				process.nextTick(() => {
+					if (lastMetadata) {
+						const includedInRange = this.lines.some(n => n >= lastCallsiteLineNumber);
+						lastMetadata.exclusive = this.match.length === 0 ?
+							includedInRange :
+							includedInRange && lastMetadata.exclusive;
+					}
+
 					hasStarted = true;
 					this.start();
 				});
 			}
 
+			if (this.lines.length > 0) {
+				this.runOnlyExclusive = true;
+				let callsiteLineNumber;
+
+				for (const jsCallsite of callsites()) {
+					const callsite = sourceMapSupport.wrapCallSite(jsCallsite);
+					const fileName = callsite.getFileName();
+					if (fileName === this.file) {
+						callsiteLineNumber = callsite.getLineNumber();
+						break;
+					}
+				}
+
+				if (lastMetadata) {
+					// `callsiteLineNumber` is the beginning of the current test, -1 to
+					// get the last line of the previous test.
+					const range = [lastCallsiteLineNumber, callsiteLineNumber - 1];
+					const includedInRange = this.lines.some(n => n >= range[0] && n <= range[1]);
+					lastMetadata.exclusive = this.match.length === 0 ?
```

Comment:
`exclusive` will be `true` if the test is declared using `.only()`. But, crucially, tests can also be skipped.

I think the ranges should narrow down from the initial test selection. If there are exclusive tests, then only those that match the ranges should be run. If there are no exclusive tests, then skipped tests should never be run.